### PR TITLE
Add GraphQL adapter for WebhooksManager

### PR DESCRIPTION
**Pull request by kirillplatonov**
### What this PR does

This PR introduces a GraphQL adapter for webhooks. It allows switching between REST and GraphQL for webhooks creation/deletion via config setting:
```ruby
config.webhook_api_adapter = :rest # default
config.webhook_api_adapter = :graphql
```
The API for webhooks management stays the same, only communication with Shopify API changes.
```ruby
config.webhooks = [
  { topic: 'app_subscriptions/update', address: "https://#{app_host}/webhooks/app_subscriptions_update" },
  { topic: 'app/uninstalled', address: "https://#{app_host}/webhooks/app_uninstalled" },
]
```

### Why do we need GraphQL for webhooks?

Unfortunately, not all webhook topics are available in REST API. For example, the topic `app_subscriptions/update` is available only in GraphQL and according to Shopify Staff replies in community threads this not going to change anytime soon:
https://community.shopify.com/c/Shopify-APIs-SDKs/webhook-for-the-topic-quot-appsubscriptions-update-quot-failing/td-p/562173
https://community.shopify.com/c/Shopify-APIs-SDKs/app-subscription-update-not-listed-anymore/td-p/581015

Since `app_subscriptions/update` is a very important webhook for any paid app I think it's a good idea to support GraphQL webhooks in this gem.

This should also resolve this issue: #1016 

### Reviewer's guide to testing

You will need a dev shop and rails app with shopify_app installed to test it.

1. Switch to graphql adapter and register `app_subscriptions/update` webhook.
2. Trigger webhook creation via following code:
```ruby
Shop.find_each do |shop|
  webhooks = ShopifyApp.configuration.webhooks
  adapter = ShopifyApp.configuration.webhook_api_adapter

  shop.with_shopify_session do
    manager = ShopifyApp::WebhooksManager.new(webhooks, adapter: adapter)
    manager.create_webhooks
  end
end
```
3. Check in graphql that webhook was added.
4. Trigger webhook destroy via following code:
```ruby
Shop.find_each do |shop|
  webhooks = ShopifyApp.configuration.webhooks
  adapter = ShopifyApp.configuration.webhook_api_adapter

  shop.with_shopify_session do
    manager = ShopifyApp::WebhooksManager.new(webhooks, adapter: adapter)
    manager.destroy_webhooks
  end
end
```
5. Check in graphql that webhook was remove.

I'm using this code in production already and it works smoothly.

### Checklist

Before submitting the PR, please consider if any of the following are needed:
- [x] Update any relevant pages in `/docs`, if necessary

### CODE REVIEW

**Comment by kirillplatonov** [comment]
> Shopify's style guide prefers allow-list/deny-list vs. white-list/black-list.

I'm not using any of these words in code. Shopify GraphQL schema contains `whitelist` word that triggered that check.

**Comment by NabeelAhsen** [comment]
Hey @kirillplatonov , have you considered suggesting this change (if this capability doesn't already exist) to `ShopifyApi`? https://github.com/Shopify/shopify_api

**Comment by kirillplatonov** [comment]
@NabeelAhsen Any chance that it can be merged? If not I can just extract this feature into a separate gem.

**Comment by lonroth** [comment]
Given that many webhooks are not available for REST but only on GraphQL make this PR very valuable and a great update to the `WebhooksManager`. 👏 Who can approve if not @awd so we can get this one merged? @kirillplatonov did you see @awd comments so the suggested changes can be discussed? I think especially `includeFields` would be great to add.

File: lib/shopify_app/adapters/webhooks/graphql_adapter.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+      def webhook_exists?(topic)
+        current_webhooks.any? do |webhook|
+          webhook["topic"] == topic
+        end
+      end
+
+      def required_webhook?(webhook)
+        webhook_address = webhook.dig("endpoint", "callbackUrl")
+        required_webhooks.any? { |w| w[:address] == webhook_address }
```
[the remark this answers is not in the document]
Comment (lonroth) [reply lib/shopify_app/adapters/webhooks/graphql_adapter.rb:95]: @awd A thing to note regarding comparing `topics` is that the topics themselves are not unique. We can register the same `topic` twice if we use two different `addresses`:
```
        {
            "id": 1028222681294,
            "address": "https://firstdomain.ngrok.io/webhooks/app_uninstalled",
            "topic": "app/uninstalled",
            "created_at": "2021-06-06T16:22:49+02:00",
            ...
        },
        {
            "id": 1028223598798,
            "address": "https://seconddomain.ngrok.io/webhooks/app_uninstalled",
            "topic": "app/uninstalled",
            "created_at": "2021-06-06T16:32:12+02:00",
            ...
        }
```
If we try to add the same address twice for the same topic we get `address for this topic has already been taken` error:
```
{
    "errors": {
        "address": [
            "for this topic has already been taken"
        ]
    }
}
```

I believe this is the reason the current codebase compares the `address` uniqueness.

File: lib/shopify_app/adapters/webhooks/graphql_adapter.rb [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      end
+
+      def create_webhook(attributes)
+        topic = graphql_topic(attributes[:topic])
+        address = attributes[:address]
```
[the remark this answers is not in the document]
Comment (lonroth) [reply lib/shopify_app/adapters/webhooks/graphql_adapter.rb:100]: I agree with @awd . This would be a really good addition. 👍

**Comment by kirillplatonov** [comment]
@awd @lonroth Thanks for the review, guys! Actually, I decided to extract these changes into a separate GraphQL client for Shopify API. The main motivation for this was the lack of any wrapper for GraphQL in ShopifyAPI (eg response wrappers, errors handling, retries, etc) and the memory issue with `shopify_api` gem that won't be resolved anytime soon (https://github.com/Shopify/shopify_api/issues/511).

I will open-source the gem soon after testing it in my app on production.

**Comment by kirillplatonov** [comment]
Closing in favor of `ShopifyGraphql` gem: https://github.com/kirillplatonov/shopify_graphql
It offers Graphql webhooks out of the box.

**Pull request closed without merging**